File: dotnet/Library/DIDComm.Messaging.gRPC/DIDComm.cs
using Google.Protobuf;

namespace DIDComm.Messaging
{
    public static class DIDComm
    {
        internal delegate int NativeMethod(ByteBuffer request, out ByteBuffer response, out ExternError error);

        internal static TResponse Call<TRequest, TResponse>(TRequest request, NativeMethod nativeMethod)
            where TRequest: IMessage
            where TResponse: IMessage, new()
        {
            using var memory = new UnmanagedMemory();

            var code = nativeMethod(memory.ToByteBuffer(request.ToByteArray()), out var response, out var error);
            memory.ThrowOnError(error);

            var res = new TResponse();
            res.MergeFrom(memory.ToArray(response));
            return res;
        }

        public static PackResponse Pack(PackRequest request) =>
            Call<PackRequest, PackResponse>(request, NativeMethods.didcomm_pack);

        public static UnpackResponse Unpack(UnpackRequest request) =>
            Call<UnpackRequest, UnpackResponse>(request, NativeMethods.didcomm_unpack);

        public static SignResponse Sign(SignRequest request) =>
            Call<SignRequest, SignResponse>(request, NativeMethods.didcomm_sign);

        public static VerifyResponse Verify(VerifyRequest request) =>
            Call<VerifyRequest, VerifyResponse>(request, NativeMethods.didcomm_verify);
    }
}

File: java/src/DIDComm/Messaging/gRPC/DIDComm.java
package DIDComm.Messaging.gRPC;

import DIDComm.Messaging.Proto.API;
import com.google.protobuf.InvalidProtocolBufferException;

public class DIDComm {
    public static API.PackResponse Pack(API.PackRequest request) throws InvalidProtocolBufferException {
        return API.PackResponse.parseFrom(NativeMethods.didcomm_pack(request.toByteArray()));
    }

    public static API.UnpackResponse Unpack(API.UnpackRequest request) throws InvalidProtocolBufferException {
        return API.UnpackResponse.parseFrom(NativeMethods.didcomm_unpack(request.toByteArray()));
    }

    public static API.SignResponse Sign(API.SignRequest request) throws InvalidProtocolBufferException {
        return API.SignResponse.parseFrom(NativeMethods.didcomm_sign(request.toByteArray()));
    }

    public static API.VerifyResponse Verify(API.VerifyRequest request) throws InvalidProtocolBufferException {
        return API.VerifyResponse.parseFrom(NativeMethods.didcomm_verify(request.toByteArray()));
    }
}
